img.js

```diff
@@ -122,8 +122,32 @@ function getValidWidths(originalWidth, widths = [], allowUpscale = false) {
   return filtered.sort((a, b) => a - b);
 }
 
+function getFileHash(src, options) {
+  const hash = createHash("sha1");
+
+  const opts = {
+    sharpOptions: options.sharpOptions,
+    sharpWebpOptions: options.sharpWebpOptions,
+    sharpPngOptions: options.sharpPngOptions,
+    sharpJpegOptions: options.sharpJpegOptions,
+    sharpAvifOptions: options.sharpAvifOptions
+  };
+
+  if(fs.existsSync(src)) {
+    const fileContent = fs.readFileSync(src);
```

Comment:
Lots of new `reads` here—I know the performance boost here is going to be a net-win but I’m curious if we can combine any?

Reply:
> if we can combine any?

Do you mean combining the `fs.existsSync` and `fs.readFileSync` operations?

If not, I'm not sure how to combine `fs.readFileSync` for multiple files, and it's probably not possible too, because `getFileHash()` knows the src only one file at a time.

But we can certainly memoize the hashes of files! We can save the hashes to a file along with the last modification time of image file, and reuse the hash if the modification time has not changed.

Reply:
Ah sorry, I mean to say that we’re reading the same file contents here at least twice—right?

* resizeImage -> getFullStats -> getStats *(runs once for each output format)* -> getFilename -> getFileHash -> `const fileContent = fs.readFileSync(src);`
* resizeImage -> `stat.buffer = fs.readFileSync(src)`

So even if we memoize the hashing (we should!) it would reduce the reads from `n+1` to `2` but I’d like to get it down to `1` 😅

Maybe we can reuse the read contents in the dryRun buffer since it’ll already have read the contents inside `getFileHash` at that point

Reply:
That's a great idea! I'll implement that, test it, and tag you when it's ready :D

Reply:
Heyo @zachleat!

So turns out we don't really need to cache `fs.readFileSync()` because the operating system already stores recently used files in memory. So repeated calls to the function are pretty much free. I used a very old and slow usb 2.0 drive to verify, and it performed identically to an nvme-ssd.

27a1a2b memoizes hashes in-memory.

I have also added code to save hashes calculated to filesystem in 65fa2f0. Hashes are stored in `.image-hash-cache.js` file inside node_modules, so the file isn't accidentally committed to git, and doesn't cause confusion to users. Memoized hashes are used if the image's last modified time is same in cache and filesystem.